activesupport/lib/active_support/callbacks.rb

```diff
@@ -642,11 +643,21 @@ def initialize_copy(other)
           @mutex     = Mutex.new
         end
 
-        def compile
-          @callbacks || @mutex.synchronize do
-            final_sequence = CallbackSequence.new
-            @callbacks ||= @chain.reverse.inject(final_sequence) do |callback_sequence, callback|
-              callback.apply callback_sequence
+        def compile(types)
+          if types.nil?
+            @callbacks || @mutex.synchronize do
+              final_sequence = CallbackSequence.new
+              @callbacks ||= @chain.reverse.inject(final_sequence) do |callback_sequence, callback|
+                callback.apply callback_sequence
+              end
+            end
+          else
+            types = Array(types)
+            @mutex.synchronize do
```

Comment:
Should we be trying to cache these here to avoid need to grab this mutex repeatedly? Gut instincts tell me yes but want to confirm

Reply:
The problem isn't really the mutex, but building the callback sequence every time. We should cache, yes.

Reply:
If it comes down to that, you can use double checked locking like on the main path.

But it would be better if the implementation didn't need to compile a distinct sequence for each variation of the `types` argument, as it would get pretty hairy and open up leak issues.

It's been a while since I dug in this code, but you didn't find a way to skip callbacks dynamically?

Reply:
@byroot My original plan was to conditionally skip callbacks in run_callbacks so we would not need to modify the sequence but it started getting a bit ugly. Doing the before/after isn't too bad but once we get in to the around callbacks it got a lot more intrusive having these checks. I found modifying the sequence itself to be the most intuitive way to approach this.